Feature/plain selenium build lib

cbeyer42 opened the pull request:
This PR introduces the plain selenium feature to run e2e tests without zalenium.

resolves #7

ppxl committed:
#7 Improve hub status check

The Selenium documentation tells us (if we want to check if hub
is available) that we should consult the `value.ready` object rather
than waiting for `status` to be zero.

The status indicates the plain availability of the hub while `ready`
indicates if the hub has actual capacity to start any tests, that is
if any are readily available.

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

ppxl committed:
Support DNS names for generated Job names

Job names are often used to create job-unique identifiers. In turn these
identifiers do also serve for docker container names which may lead to DNS
names. DNS names should not contain underscores.

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

ppxl committed:
WIP

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

cbeyer42 committed:
#7 finish removal of selenium containers

cbeyer42 committed:
#7 update default values and some variable names

cbeyer42 committed:
#7 add debug logging

cbeyer42 committed:
#7 pass network name to shared selenium lib

cbeyer42 committed:
#7 add method for generating selenium job name

cbeyer42 committed:
#7 add check for docker network

cbeyer42 committed:
#7 resolve issue with undefined variable

cbeyer42 committed:
#7 pass arguments when starting docker container

cbeyer42 committed:
#7 use right method signature

cbeyer42 committed:
#7 remove network after closure execution is done

cbeyer42 committed:
#7 remove docker network if it exists

cbeyer42 committed:
#7 remove redundant rm of grid hub

cbeyer42 committed:
#7 change the comments to selenium instead of zalenium

cbeyer42 committed:
#7 remove own network handling and some dead code

The docker network should only be specified when it exists. Otherwise the build step will fail.

cbeyer42 committed:
#7 use network for selenium hub and node containers

cbeyer42 committed:
#7 create default network for selenium communication

If there is no network passed to selenium a default network will be created. This network is used to establish the communication between selenium grid hub and its nodes with the container name of the grid hub.

cbeyer42 committed:
#7 remove static modifiers

cbeyer42 committed:
#7 refactor network creation

cbeyer42 committed:
#7 maybe fix an error on script execution

cbeyer42 committed:
#7 print error message when network creating failed

cbeyer42 committed:
#7 don't fail the build job when an error occures during list of networks

cbeyer42 committed:
#7 move network creation out of the selenium hub container scope

cbeyer42 committed:
#7 another try to fix the error problem

cbeyer42 committed:
#7 another try to fix it (2)

cbeyer42 committed:
#7 finally fix it

cbeyer42 committed:
#7 fix type in network variable name and modify helper

cbeyer42 committed:
#7 remove unused function from helper

cbeyer42 committed:
#7 fix wrong argument count

cbeyer42 committed:
#7 revert deleting helper script

cbeyer42 committed:
#7 add the ability to customize the maven central mirror path

Previously the path was hard coded.

cbeyer42 committed:
#7 refactor network check

cbeyer42 committed:
#7 remove unnecessary character for maven mirror

cbeyer42 committed:
#7 pass debug parameter to grid hub and nodes

cbeyer42 committed:
#7 add settings overload for maven-central mirror

cbeyer42 committed:
#7 simplify selenium build lib and update comments

In addition, the usage description was updated so it shows how selenium can be configured.

cbeyer42 committed:
#7 add CHANGELOG and update description file for withMavenSettings

ppxl commented on CHANGELOG.md:
```diff
@@ -0,0 +1,14 @@
+
+### Changed
+- signatures for `withMavenSettings` changed slightly so the maven central mirror can be configured
+- helper `generateZaleniumJobName` was renamed to `generateJobName`
```
Note: `generateJobName()` has changed its format because of possible problems when used as DNS name.

ppxl commented on vars/withMavenSettings.groovy:
```diff
@@ -50,7 +51,14 @@ def settings(def nexusCredentials, String cesFQDN, String pathToLocalMavenReposi
 def mvnWithSettings(def nexusCredentials, String cesFQDN, String mvnCallArgs) {
     def currentHome = env.HOME
-    settings(nexusCredentials, cesFQDN, currentHome) { settingsXml ->
+    settings(nexusCredentials, cesFQDN, currentHome, "maven-central/") { settingsXml ->
```
I don't get this `maven-central/` reference. Why would this be the default path to a repository?

cbeyer42 replied:
`maven-central` is a nexus repository which proxies the requests to the official maven central repo.

ppxl commented on vars/withMavenSettings.groovy:
```diff
@@ -60,16 +68,21 @@ def mvnWithSettings(def nexusCredentials, String cesFQDN, String mvnCallArgs) {
 def mvnWithHome(String mvnHome, String settingsXml, String mvnCallArgs) {
     sh "${mvnHome}/bin/mvn -s ${settingsXml} --batch-mode -V -U -e -Dsurefire.useFile=false ${mvnCallArgs}"
-}
+}
```
Here is a trailing newline missing

ppxl commented on vars/withMavenSettings.txt:
```diff
@@ -33,15 +33,18 @@ Example call with included Maven call:
-- settings(def nexusCredentials, String cesFQDN, String pathToLocalMavenRepository, Closure closure) -  Writes a
-  `settings.xml` file with the given Nexus credentials and executes the closure body. This method provides a parameter
-  `settingsXml` which contains the relative file path to the 'settings.xml' file.
+- settings(def nexusCredentials, String cesFQDN, String pathToLocalMavenRepository, Closure closure)
```
I see you take care of the slightest aspects :+1:

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+            debugSelenium     : false
+    ]
+
+    def networkExists = checkNetwork(seleniumNetwork)
```
This looks like a nice solution.

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+
+Boolean checkNetwork(String networkName) {
+    return sh(returnStatus: true, script: "docker network ls | grep ${networkName}") == 0
+}
```
Here went a tailing newline missing

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+    def uid = findUid()
+    def gid = findGid()
+
+    def networkParameter = ""
```
Can this be a thing? An empty network after you already asserted its existence?

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    gridDebugParameter = ""
+    if (gridDebugParameter != null && !gridDebugParameter.isEmpty()) {
```
Did you mean `debugSelenium`? I don't see any other variable from the map

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+    seleniumHubImage.pull()
+    // Run with Jenkins user, so the files created in the workspace by selenium can be deleted later
+    // Otherwise that would be root, and you know how hard it is to get rid of root-owned files.
+    def dockerArgs = "-u ${uid}:${gid} ${networkParameter} ${gridDebugParameter} -p 4444:4444 --name ${hubName}"
```
A hardcoded port on the host side? That will never lead to a problem :rofl: But I'll leave it :wink:

cbeyer42 replied:
I changed it so it can be configured. The default value is 4444.

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+String generateJobName() {
```
A large part of them methods here and downward are utility methods with no direct connection to running selenium, and I swear I saw them somewhere else too: :thinking: Could they be split off and shared among the library parts? They surely are used there too, aren't they?

cbeyer42 replied:
I tried to extract this function but I did not get it to work. Maybe the `import` and `evaluate` mechanisms work diffently for jenkins.

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+
+    def workerImage = docker.image(workerNodeImage)
+    workerImage.pull()
+    GString dockerDefaultArgs = "${networkParameter} ${debugParameter} -e HUB_HOST=${hubHost} -v /dev/shm:/dev/shm"
```
`def` instead of GString would do the trick as well

ppxl commented on vars/withSelenium.groovy:
```diff
@@ -0,0 +1,176 @@
+    GString dockerDefaultArgs = "${networkParameter} ${debugParameter} -e HUB_HOST=${hubHost} -v /dev/shm:/dev/shm"
+    ArrayList<String> workerIDList = []
+    for (int i = 0; i < count; i++) {
+        container = workerImage.run(dockerDefaultArgs)
```
Maybe log a bit of output for the docker interaction here?

ppxl commented on vars/withZalenium.groovy:
```diff
@@ -119,7 +119,7 @@ void waitForSeleniumToGetReady(String host) {
     sh(returnStdout: true,
             script: "curl -sSL http://${host}:4444/wd/hub/status || true") // Don't fail
-            .contains('status\": 0')
+            .contains('ready\": true')
```
Nice catch!

Did you check the output? Ready == true vs the hub is capable of starting new workers?

cbeyer42 replied:
Ready is only true when there are enough idle worker nodes connected to the grid hub.

cbeyer42 committed:
#7 use shared helper functions

cbeyer42 committed:
#7 update changelog

cbeyer42 committed:
#7 wrap helper functions with class

cbeyer42 committed:
#7 revert change helper functions to closures

cbeyer42 committed:
#7 add port mapping for the grid hub

cbeyer42 committed:
#7 fix usage of debug config parameter

cbeyer42 committed:
#7 re-add generateZaleniumJobName to grant backwards compatibility and update documentation

ppxl commented on vars/withMavenSettings.groovy:
```diff
@@ -15,8 +15,9 @@
  * @param nexusCredentials Jenkins credentials which provide USERNAME and PASSWORD to an account which enables Nexus interaction
  * @param cesFQDN the full qualified domain name of the current CES instance, f. i. <code>cesinstance.stage.host.tld</code>
  * @param pathToLocalMavenRepository without the .m2 directory part, f. i. <code>/usr/share/maven</code>. The suffix <code>/.m2/repository</code> will be added automatically.
+ * @param mirrorNexusPath relativ path to a maven central mirror hosted inside a nexus instance. The suffix <code>/nexus/repository</code> will be added automatically.
```
- typo: `relativ` -> e
- provide example for generated output string: https.../../m.n.p.

ppxl commented on vars/withMavenSettings.txt:
```diff
@@ -1,6 +1,7 @@
 context of a Cloudogu EcoSystem. This global variable writes custom `settings.xml` files and can additionally invoke a
-provided or custom Maven call.
+provided or custom Maven call. All calls to any of the listed maven settings functions use the in nexus integrated maven
+central proxy repository if there is no other maven central mirror configured.
```
... if no further maven central mirror is configured

ppxl reviewed (commented):
only slight items arise here, mainly non-functional things like wording of comments, or newlines. I'll fix it and than we're through.

ppxl committed:
#7 Fix typos and add some more doc comments

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

ppxl committed:
#7 Fix typos

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

ppxl committed:
#7 Add trailing newline

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

ppxl committed:
#7 Some more doc comments

Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

ppxl committed:
#7 Some more doc comments and one less breaking change

Previously the signature of `settings()` was modified to contain a maven repo
name. This would be a breaking change. This new parameter was now retrofitted
with a senseful default so it still could work out.

Furthermore: Doc-o-rama
Signed-off-by: Philipp Pixel <pixeloffice@philipppixel.de>

cbeyer42 committed:
#7 unify the usage of ArrayList

In some cases the use of `toArray` is forbidden.

cbeyer42 committed:
#7 fix wrong port binding for selenium hub

ppxl commented:
LGTM
